schema/function_test.go

```diff
@@ -0,0 +1,517 @@
+		simpleIntSchema,
+		nil,
+		func(a int64, b int64, c int64) (int64, error) {
+			return a * b * c, fmt.Errorf("this is an error")
+		},
+	)
+	assert.NoError(t, err)
+	_, err = simpleFunc.Call([]any{int64(1), int64(2), int64(3)})
+	// There should be an error. Validate that it's the correct one.
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "this is an error")
+}
+
+func TestCallableFunctionSchema_Err_TestIncorrectNumArgs(t *testing.T) {
+	simpleIntSchema := schema.NewIntSchema(nil, nil, nil)
+	simpleFunc, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{simpleIntSchema, simpleIntSchema, simpleIntSchema}, // Three params
+		simpleIntSchema,
+		nil,
+		func(a int64, b int64, c int64) (int64, error) { // Three params
+			return a * b * c, nil
+		},
+	)
+	assert.NoError(t, err)
+	_, err = simpleFunc.Call([]any{int64(1), int64(2)}) // Two args specified here
+	// There should be an error. Validate that it's the correct one.
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "incorrect number of args")
+}
+
+// The following test requires bypassing the validation provided in NewCallableFunction.
+func TestCallableFunctionSchema_Err_CallWrongErrorReturn(t *testing.T) {
+	callable := schema.CallableFunctionSchema{
+		IDValue:            "test",
+		InputsValue:        []schema.Type{},
+		DefaultOutputValue: nil, // No return specified here, so only zero returns, or an error return is allowed.
+		DisplayValue:       nil,
+		Handler: reflect.ValueOf(func() any { // Non-error return schema.Type here
+			return 5
+		}),
+	}
+	_, err := callable.Call(make([]any, 0))
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "error return val isn't an error")
+}
+
+// The following test requires bypassing the validation provided in NewCallableFunction.
+func TestCallableFunctionSchema_Err_CallWrongReturnCount(t *testing.T) {
+	callable := schema.CallableFunctionSchema{
+		IDValue:            "test",
+		InputsValue:        []schema.Type{},
+		DefaultOutputValue: nil, // No returns specified here
+		DisplayValue:       nil,
+		Handler: reflect.ValueOf(func() (any, any, any) { // Three returns specified here
+			return 0, 0, 0
+		}),
+	}
+	_, err := callable.Call(make([]any, 0))
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "unexpected return count")
+}
+
+// Test the input validation.
+func TestNewCallableFunction_Err_MismatchedParamCount(t *testing.T) {
+	_, err := schema.NewCallableFunction(
+		"test",
+		make([]schema.Type, 0), // No params specified here
+		nil,
+		nil,
+		func(int, int) {}, // Two params specified here
+	)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "parameter inputs do not match handler inputs")
+}
+
+func TestNewCallableFunction_Err_MismatchedParamType(t *testing.T) {
+	_, err := schema.NewCallableFunction(
+		"test",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)}, // String specified here
+		nil,
+		nil,
+		func(int) {}, // Int specified here, mismatched
```

Comment:
I think they make it easy for those who aren't experienced with the code base. I'll leave them in because the clutter they add isn't worse than the value they add, in Matt and I's opinions.